\documentclass{article}
\usepackage{graphicx} % Required for inserting images

\usepackage[numbers]{natbib}  % uses numbers for citations
% \usepackage[semicolon]{natbib}
\usepackage{url}
\usepackage{amsmath}
\usepackage{todonotes}


\title{CMSC702 Project}
\author{Vincent La, Spencer Jenkins, Niko Zheng, Amaan Makandar}
\date{November 2024}



\begin{document}

\maketitle

% TODO for spencer: find digital images of proteins we look at
\section{Abstract}

In this project, we evaluate the consistency between protein contact predictions generated by ESM-2, a transformer-based protein language model, and traditional coevolutionary coupling methods such as Direct Coupling Analysis (DCA). Unlike methods like AlphaFold that rely on multiple sequence alignments (MSAs), ESM-2 is trained on raw sequence data, which underlines the need to verify its consistency. We used both mean-field DCA (mfDCA) and pseudolikelihood maximization DCA (plmDCA) to infer residue-residue couplings based on coevolutionary signals in MSAs as a point of comparison to ESM structure prediction. Our preliminary results reveal limitations in resolution for smaller MSAs and highlight discrepancies between DCA and ESM-2 predictions, which may stem from differences in their underlying assumptions and data dependencies. These findings provide insight into the capabilities and constraints of ESM-2, laying the groundwork for further studies on its application to protein structure prediction.

% TODO WRITE FINAL CONCLUSIONS

\section{Introduction}

The construction of multiple sequence alignments (MSAs) of genetic sequences has long been fundamental to computational phylogenetics. However, recent advances have led many researchers in the field to move beyond traditional algorithms, drawing inspiration from breakthroughs in natural language processing (NLP) and applying neural network technologies to computational biology. In particular, neural transformer-based models, such as the protein language model ESM \cite{ESM-FOLD}, hold great promise for advancing our understanding of protein structure, function, and evolution.

\subsection{Evolutionary Scale Modeling (ESM)}
ESM is a family of protein language models developed by Meta that use a transformer-like architecture similar to BERT. The specific version used in this project, ESM-2, is pretrained using a masked language modeling approach involving predicting masked residues based on their surrounding sequence context. This pretraining makes ESM-2 highly adaptable to downstream tasks, including secondary structure prediction. In our project, we examine two downstream adaptations of ESM: ESM+Contact Head and ESM-FOLD.

\subsection{ESM+Contact Head}

To use ESM-2 for protein contact prediction, Rao et al. build upon ESM with a lightweight ``Contact head'' trained to output 2-D residue contact maps\cite{ESMContactHead}. The ``Contact Head'' is a logistic regression model trained on the attention maps from each layer of ESM-2. The approach takes heavy inspiration from language model probing, which encompasses a variety of studies into the internal representations learned by language models wherein the outputs of the hidden layers of a model are used to train lightweight classifiers. Rao et al. argue that their ``Contact Head'' demonstrates the immense potential of transformers in learning protein structure, and the ``Contact Head'' would soon be superseded by far larger and more versatile models.

\subsection{ESM-FOLD}

ESM-FOLD \cite{ESM-FOLD} is a large transformer-based model designed to predict the three-dimensional structures of protein sequences. It leverages ESM-2 as a encoder, passing the resulting embeddings to a structure module to generate structure predictions The model outputs pairwise residue distances, with those below a specified distance threshold classified as being in "contact." ESM-FOLD is trained in a supervised manner with sequences and experimentally determined structures. Unlike the Contact Head, ESM-FOLD outputs an atomic-scale protein structure in addition to the protein contact map. Both the Contact Head and ESM-FOLD have been used together in comparative studies in the past, to determine what, if any, insight about protein structure is learned by ESM-2. \cite{Zhang}

\subsection{Predecessors of ESM}

A key innovation of ESM lies in its training approach: it exclusively requires raw sequence data as input. This stands in contrast to AlphaFold\cite{AlphaFold}, the Nobel Prize-winning model that can also predict residue-residue contact maps and relies on database searches for evolutionary information (MSAs) and structures for its input. Given this distinction, we aim to explore the capabilities of ESM-2, particularly in its ability to infer meaningful residue contacts without the use of MSAs. Our objective is to evaluate the trustworthiness of ESM-2 by assessing its consistency with established methods for contact prediction. Studying both ESM+Contact Head and ESM-FOLD gives us a greater picture of the insight that ESM-2 gains from pretraining on its own. 

\subsection{Direct Coupling Analysis (DCA)}
We use direct coupling analysis to evaluate the consistency of both protein language models. DCA is a statistical framework used to predict pairwise relationships between residue sites in a protein sequence based on the coevolutionary signal encoded in multiple sequence alignments (MSAs). The method models the joint probability distribution \( P(\mathbf{x}) \) of a protein sequence alignment, where \( \mathbf{x} = (x_1, x_2, \ldots, x_L) \) represents the amino acid residues at \( L \) sites in the sequence. DCA approximates \( P(\mathbf{x}) \) using a pairwise maximum entropy model, the simplest model that satisfies the observed single-site frequencies \( f_i(x_i) \) and pairwise frequencies \( f_{ij}(x_i, x_j) \) derived from the MSA:  
\[
P(\mathbf{x}) \propto \exp\left(\sum_{i} h_i(x_i) + \sum_{i<j} J_{ij}(x_i, x_j)\right),
\]  
where \( h_i(x_i) \) are site-specific bias parameters, and \( J_{ij}(x_i, x_j) \) are coupling parameters quantifying the statistical dependency (or coevolution) between residue pairs \( i \) and \( j \). The magnitudes of the coupling parameters \( J_{ij} \) are interpreted as indicators of direct interactions, allowing DCA to identify residue-residue contacts critical for protein folding and structural analysis. 

\subsection{Approximations of DCA}

Calculating the exact parameters \( h_i \) and \( J_{ij} \) involves solving a highly intractable optimization problem due to the exponential complexity of the sequence space. Therefore, practical implementations rely on approximations. We evaluate two approximations of DCA in our study: mean-field and pseudolikelihood maximization. Mean-field direct coupling analysis\cite{Morcos} (mfDCA) approximates the effect of all other residues on a given pair as an average or ``mean field''. The mean-field approximation is less accurate than other methods, in that it does a poor job at predicting coupling between sectors as opposed to between individual residues. However, it is highly efficient, running up to 10,000 times faster than its predecessors.

Pseudolikelihood maximization direct coupling analysis (plmDCA)\cite{ekeberg} approximates the full joint probability distribution for the residues by calculating conditional probabilities of each residue given all others using the maximum entropy model. The plmDCA method strikes a balance between computational efficiency and accuracy, as it is not as efficient as mfDCA but more accurate at capturing direct dependency relationships.

\subsection{Consistency Evaluation}

We use these two methods, along with publicly available experimental protein structure determinations, to assess the accuracy of ESM-2 contact predictions. Because residues in contact undergo evolutionary pressure to coevolve, we expect to see agreement between DCA and ESM-FOLD. We aim to quantify the consistency between DCA and ESM-FOLD by using the Spearman correlation, which considers rank as well as absolute value, to quantify the differences between predicted values for each residue pair for a given protein. Furthermore, we use ``precision@K'' and ``recall@k'', where precision indicates the fraction of predictions from a given model being correct, and recall indicates the fraction of contact sites being correctly identified. The ``@k'' represents that the top $k$ scores from the model are taken. We vary $k$ to determine the threshold at which sites are predicted correctly. We detail these metrics further in part 3.4. 

We are studying both ESM+Contact Head and ESM-FOLD because we believe that the ``Contact Head'' is more representative of what ESM-2 learns in pretraining compared to ESM-FOLD which is supervised on protein structures. We expect to see more consistency between ESM-FOLD and DCA for this reason.

\section{Methods}

To measure the consistency between DCA scores and ESM contact predictions, we analyze popular protein families such as Cadherin and Casein, both of which have publicly available MSAs from the Pfam database. We ran inferences with ESM+Contact Head and ESM-FOLD to get the contact prediction scores for all residual pairs and compared it to the mfDCA and plmDCA scores using histograms, heatmaps, and the Spearman coefficient, as well as precision@k and recall@k. 

It is important to note that in secondary structure prediction diagrams, it is common to see a strong diagonal line from the top left to the bottom right, indicating high likelihood or degree of contact with immediate neighbor residues. Although this diagonal axis may appear to predict contact between proteins that are already close to each other, it often indicates the presence of $\alpha$-helices or forward $\beta$-sheets. To avoid giving undue weight to such contact predictions, we analyzed contacts for residue pairs that are at least 4 sites apart on the sequence.

The next section gives details on how we preprocessed the datasets for the study. We first outline the protein data we are using, and then outline the software libraries we are using.

\subsection{Cadherin PF00028 dataset}
The first protein family we studied was Cadherin PF00028 MSA from Pfam, which consists of over 500 thousand sequences. Because we could not run ESM-FOLD and DCA on the entire MSA due to its large size, we opted to run them on only 2000 random sequences, given that previous DCA studies use up to 1000 random sequences.

\begin{table}[]
    \centering
    \resizebox{\textwidth}{!}{
    \begin{tabular}{|c|c|c|c|c|c|}
    \hline
          Pfam ID & PBD Name & PDB ID & \multicolumn{1}{|p{2cm}|}{\centering Uniprot Residue Indices (full)} & \multicolumn{1}{|p{2cm}|}{\centering Uniprot Residue Indices (subset)} & B\_eff \\ \hline
          PF00076 & ELAV3\_MOUSE & 1d8z & [35,123] & [41-111] & 4561.72 \\ \hline
          PF00028 & CADH1\_HUMAN & 2o72 & [155,367] & [267,366] & 5933.55 \\ \hline
          PF00011 & CRYAA\_BOVIN & 3L1E & [59, 163] & [63-162] & 6594.44 \\ \hline
          PF00043 &  &  &  &  & \\
    \hline
    \end{tabular}
    }
    \caption{The table above describes the MSA datasets used for our study along with specific preprocessing details.}
    \label{tab:my_label}
\end{table}

\subsection{Mean field direct coupling analysis}
To get the mfDCA scores, we used the publicly available Github repository py-mfdca owned by utdal\cite{pydca}, which is a Python port of the MATLAB library included with the original mfDCA paper. 

\subsection{Pseudolikelihood maximization direct coupling analysis}
We used a C-based library for computing the plmDCA scores.  

\subsection{Metrics}
To evaluate the performance of ESM-2 and DCA contact prediction, we compute the Spearman correlation, a measure of correlation that incorporates rank as well as numerical similarity. 

We also draw from information theory and report precision and recall between the methods we want to demonstrate consistency for. We compute precision@$k$ which we define as:

$$\text{precision@}k = \frac{\text{Number of true contacts in top-k predictions/scores}}{k} $$

\noindent where true contacts are determined from a PDB structure file that is representative of the MSA of interest. When $k$ is equal to $L$, the length of the sequence, we have a metric that evaluates the general performance of the contact prediction method. Since this is a ratio, precision@$k$ is bounded between $\left[0,1\right]$.

We also compute recall@$k$ between ESM-2 contact predictions and DCA scores which we define as:

$$\text{recall@}k = \frac{\text{Number of coupled pairs in top-k ESM predictions}}{\text{Total true DCA couplings}} $$

These metrics are not sensitive to ranking, unlike Spearman correlation. 

We also consider qualitative metrics, such as the appearance of scatter plots of the predicted residue contacts. Two types of plots we present in this paper are residue distance vs. contact score, and 


\noindent 
\section{Results}

We studied the consistency between the metrics by creating correlation scatter plots between mfDCA, plmDCA, and ESM.

The preliminary DCA score distributions for each method reveal a normal distribution of scores with a mean of around 0 and a standard deviation of around 0.01 for plmDCA, and an irregular distribution for mfDCA with only positive scores with a mode of 0 and trailing to the right. 

We also created a plot comparing the residue pair distance as experimentally determined 

\section{Discussion}

In our preliminary study, the heat maps we obtained for both DCA methods had vertical and horizontal streaks. This was likely a sign of low signal resolution, due to the cadherin MSA consisting of only 55 sequences. Typically, DCA is used with MSAs in the order of thousands of sequences. Accordingly, there was little we could interpret from our preliminary correlation scatter plots, which showed little to no correlation between ESM and either DCA method. Furthermore, the Spearman correlation distribution plots revealed very poor correlation between ESM and both DCA methods. 



\section{Conclusion}



\section{Keywords/references to be added in references section TODO}
Alphafold \cite{AlphaFold}
pydca \cite{pydca}
PL-DCA \cite{Ekeberg}
Mean-field DCA \cite{Morcos}
pLM categorical-jacobian study \cite{Zhang}
ESM w/ contact head \cite{ESMContactHead}


\bibliographystyle{unsrtnat} % orders by order of citatation in file
\bibliography{refs}

\end{document}
